pkg/ebpf/bpf_errors.go

```diff
@@ -0,0 +1,228 @@
+package ebpf
+
+import (
+	"encoding/binary"
+	"fmt"
+	"math"
+	"strings"
+	"unsafe"
+
+	"github.com/aquasecurity/tracee/pkg/logger"
+)
+
+const BPF_MAX_ERR_FUNC_LEN = 48
+
+type BPFErrorType uint32
+
+const (
+	BPF_ERR_MAP_UPDATE_ELEM BPFErrorType = iota + 1
+	BPF_ERR_MAP_DELETE_ELEM
+	BPF_ERR_GET_CURRENT_COMM
+	BPF_ERR_TAIL_CALL
+
+	BPF_ERR_UNSPEC BPFErrorType = math.MaxUint32
+)
+
+var stringMap = map[BPFErrorType]string{
+	BPF_ERR_MAP_UPDATE_ELEM:  "BPF_ERR_MAP_UPDATE_ELEM",
+	BPF_ERR_MAP_DELETE_ELEM:  "BPF_ERR_MAP_DELETE_ELEM",
+	BPF_ERR_GET_CURRENT_COMM: "BPF_ERR_GET_CURRENT_COMM",
+	BPF_ERR_TAIL_CALL:        "BPF_ERR_TAIL_CALL",
+
+	BPF_ERR_UNSPEC: "BPF_ERR_UNSPEC",
+}
+
+func (b BPFErrorType) String() string {
+	if bpfErr, found := stringMap[b]; found {
+		return bpfErr
+	}
+
+	return stringMap[BPF_ERR_UNSPEC]
+}
+
+var errorMap = map[BPFErrorType]string{
+	BPF_ERR_MAP_UPDATE_ELEM:  "Failed to add or update a map element",
+	BPF_ERR_MAP_DELETE_ELEM:  "Failed to delete a map element",
+	BPF_ERR_GET_CURRENT_COMM: "Failed to get current command",
+	BPF_ERR_TAIL_CALL:        "Failed to tail call",
+
+	BPF_ERR_UNSPEC: "Unspecifed BPF error",
+}
+
+func (b BPFError) Error() string {
+	if bpfErr, found := errorMap[BPFErrorType(b.id)]; found {
+		return bpfErr
+	}
+
+	return errorMap[BPF_ERR_UNSPEC]
+}
+
+type BPFError struct {
+	ret      uint64
+	id       uint32
+	line     uint32
+	function [BPF_MAX_ERR_FUNC_LEN]byte
+}
+
+func NewBPFError() BPFError {
+	return BPFError{
+		id: uint32(BPF_ERR_UNSPEC),
+	}
+}
+
+func (b BPFError) Return() int64 {
+	return int64(b.ret)
+}
+
+func (b BPFError) ID() uint32 {
+	return uint32(b.id)
+}
+
+func (b BPFError) Type() BPFErrorType {
+	return BPFErrorType(b.id)
+}
+
+func (b BPFError) Line() uint32 {
+	return b.line
+}
+
+func (b BPFError) Function() []byte {
+	return b.function[:]
+}
+
+func (b BPFError) FunctionAsString() string {
+	errFunc := string(b.function[:])
+	errFuncEnd := strings.Index(errFunc, "\x00")
+	return errFunc[:errFuncEnd]
+}
+
+func (b BPFError) returnSize() int {
+	return int(unsafe.Sizeof(b.ret))
+}
+
+func (b BPFError) idSize() int {
+	return int(unsafe.Sizeof(b.id))
+}
+
+func (b BPFError) lineSize() int {
+	return int(unsafe.Sizeof(b.line))
+}
+
+func (b BPFError) functionSize() int {
+	return int(len(b.function))
+}
+
+func (b BPFError) Size() int {
```

Comment:
I suppose that Size() returns a compile-time calculated constant, but I need to be sure.

Reply:
Because it uses dynamic unsafe.Sizeof checks i'm not sure it does but either way we may see bottlenecks in other areas (and this part may be insignificant), this just put the idea in my mind.

Reply:
> Because it uses dynamic unsafe.Sizeof checks i'm not sure it does but either way we may see bottlenecks in other areas (and this part may be insignificant), this just put the idea in my mind.

I got your point and thought about that as well.

It seems that the compiler takes care and calculates the return for us: https://go.godbolt.org/z/Kc3jbWanM

```asm
        TEXT    main.BPFError.Size(SB), NOSPLIT|ABIInternal, $0-64
        FUNCDATA        $0, gclocals·g2BeySu+wFnoycgXfElmcg==(SB)
        FUNCDATA        $1, gclocals·g2BeySu+wFnoycgXfElmcg==(SB)
        FUNCDATA        $5, main.BPFError.Size.arginfo1(SB)
        MOVL    $64, AX
        RET
        TEXT    main.BPFError.SizeStatic(SB), NOSPLIT|ABIInternal, $0-64
        FUNCDATA        $0, gclocals·g2BeySu+wFnoycgXfElmcg==(SB)
        FUNCDATA        $1, gclocals·g2BeySu+wFnoycgXfElmcg==(SB)
        FUNCDATA        $5, main.BPFError.SizeStatic.arginfo1(SB)
        MOVL    $64, AX
        RET
```